src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,493 @@
+            + "\nThe waterway crosses a coastline, which means it is possible for the coastline to have an incorrect direction.\nLand should be to the LEFT of the coastline and the ocean should be to the RIGHT of the coastline (for more information, see https://wiki.osm.org/Tag:natural=coastline).";
+    private static final String CIRCULAR_WATERWAY = "The waterway {0} loops back on itself. This is typically impossible.";
+    private static final String CROSSES_WATERWAY = "The waterway {0} crosses the waterway {1}.";
+    private static final String GOES_UPHILL = "The waterway {0} probably does not go up hill.\nPlease check (source elevation data resolution was about {1} meters).";
+    private static final Distance MIN_RESOLUTION_DISTANCE = Distance.ONE_METER;
+
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays.asList(DOES_NOT_END_IN_SINK,
+            CIRCULAR_WATERWAY, CROSSES_WATERWAY, GOES_UPHILL,
+            DOES_NOT_END_IN_SINK_BUT_CROSSING_OCEAN);
+
+    private final TaggableFilter waterwaySinkTagFilter;
+    private final TaggableFilter waterwayTagFilter;
+
+    private final TaggableFilter validOceanTags;
+
+    private final TaggableFilter oceanBoundaryTags;
+    private final ElevationUtilities elevationUtils;
+    private final Distance minResolutionDistance;
+    private final Distance minDistanceStartEndElevationUphill;
+
+    /**
+     * Get intersecting segments of two polylines
+     *
+     * @param left
+     *            The polyline whose segments will go in the key/left side of the pair
+     * @param right
+     *            The polyline whose segments will go in the value/right side of the pair
+     * @return A collection of segments intersecting other segments. A segment *may* appear more
+     *         than once.
+     */
+    public static Collection<Pair<Segment, Segment>> getIntersectingSegments(final PolyLine left,
+            final PolyLine right)
+    {
+        final Collection<Pair<Segment, Segment>> intersectingSegments = new HashSet<>();
+        for (final Segment leftSegment : left.segments())
+        {
+            for (final Segment rightSegment : right.segments())
+            {
+                if (leftSegment.intersects(rightSegment))
+                {
+                    intersectingSegments.add(Pair.of(leftSegment, rightSegment));
+                }
+            }
+        }
+        return intersectingSegments;
+    }
+
+    /**
+     * Check the atlas object for boundary nodes at its end
+     *
+     * @param object
+     *            The atlas object to check
+     * @return {@code true} if the object ends on a boundary
+     */
+    private static boolean endsWithBoundaryNode(final AtlasObject object)
+    {
+        if (!(object instanceof LineItem))
+        {
+            return false;
+        }
+        final LineItem lineItem = (LineItem) object;
+        final Atlas atlas = object.getAtlas();
```

Comment:
`getAtlas()` can be an expensive operation. You should be able to store the atlas in the flag function in pass it to your helpers that need access to the entire atlas.